Explain the bridge between SF and legacy hooks (#661)
PrestaShop/docs: src/content/1.7/development/components/hooks/symfony-bridge.md

---
title: Symfony bridge for hooks
---

# Symfony bridge for hooks

In Back-Office, migrated pages are powered by Symfony. The use of the powerful [EventDispatcher][sf-event-dispatcher] of Symfony has been chosen to replace, in the long term, hooks. However for now both systems are used in order to preserve backward compatibility.

In Symfony, there is a [HookDispatcher][sf-hook-dispatcher-class] class that actually uses the dispatch system of Symfony.

It is important that modules which subscribe to a hook X are correctly triggered, whether the hook is being dispatched in legacy code or in modern code.

In legacy code, the class Hook acts as usual.

In Symfony-powered code, multiple classes work together to achieve this.

The class [LegacyHookSubscriber][legacy-hook-subscriber] acts as a registry for legacy hooks and gateway between legacy and Symfony.

Class HookDispatcher is actually a wrapper of Symfony EventDispatcher.

## Registry

First, on setup, class LegacyHookSubscriber will retrieve all legacy hooks using `hooks = Hook::getHooks();`.

Then it will create as many event subscriptions as there are hooks, following a naming convention based on database IDs.

> For example if module with `id_module` 267 has subscribed to hook with `id_hook` 82, `LegacyHookSubscriber` will create a subscription `call_82_267`.

Reviewer: ```suggestion
For example, if module with `id_module` 267 has subscribed to hook with `id_hook` 82, `LegacyHookSubscriber` will create a subscription `call_82_267`.
```

## Dispatching

In Symfony-powered pages, to dispatch hook one must use the HookDispatcher:

```php
$this->hookDispatcher->dispatchWithParameters("actionModifyForm", ['form_builder' => $formBuilder]);
```

The HookDispatcher will act as standard dispatcher and call the eligible event listeners and subscribers, including LegacyHookSubscriber.

Using a dedicated magic method `__call()`, the `LegacyHookSubscriber` will parse the called method (ex: `call_82_267`) to retrieve the appropriate hook and module (using their respective IDs) and trigger the related `Hook::exec()` call with the appropriate parameters.

## To sum up

> When a hook is dispatched inside Symfony-powered pages of the Back-office, the HookDispatcher dispatches the hook similarly to a regular Symfony event, and this event is being listened to by LegacyHookSubscriber that will then trigger the related `Hook::exec()` call.

Reviewer: ```suggestion
When a hook is dispatched inside Symfony-powered controllers, `HookDispatcher` dispatches the hook similarly to a regular Symfony event; this event is being listened to by `LegacyHookSubscriber` who will then trigger the related `Hook::exec()` call.
```
